umh-core/pkg/fsm/agent/reconcile.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2025 UMH Systems GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	internal_fsm "github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/internal/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/backoff"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/metrics"
+)
+
+// Constants for agent FSM
+const (
+	// AgentUpdateObservedStateTimeout defines the timeout for updating observed state
+	AgentUpdateObservedStateTimeout = time.Millisecond * 5
+)
+
+// Reconcile examines the AgentInstance and, in three steps:
+//  1. Check if a previous transition failed; if so, verify whether the backoff has elapsed.
+//  2. Detect any external changes (e.g., a new configuration or external signals).
+//  3. Attempt the required state transition by sending the appropriate event.
+//
+// This function is intended to be called repeatedly (e.g. in a periodic control loop).
+func (a *AgentInstance) Reconcile(ctx context.Context, tick uint64) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(ComponentAgentInstance, "reconcile", time.Since(start))
+		if err != nil {
+			a.baseFSMInstance.GetLogger().Errorf("Error reconciling agent: %v", err)
+			a.PrintState()
+			// Add metrics for error
+			metrics.IncErrorCount(ComponentAgentInstance, "agent")
```

Comment:
why did you choose "agent"? if there is only one, then probably use "Core" as this is how we do it in other modules